Coap options random order

LukasWoodtli opened the pull request:
Allow the insertion of an CoAP option in between of existing options. All options following the new option (and also any payload) are moved in memory. The single option directly following the new option is rewritten (after removal) as the code offset changes.

rlubos commented on subsys/net/lib/coap/coap.c:
```diff
@@ -49,53 +49,132 @@ LOG_MODULE_REGISTER(net_coap, CONFIG_COAP_LOG_LEVEL);
 static uint16_t message_id;
 
-static inline bool append_u8(struct coap_packet *cpkt, uint8_t data)
+static int parse_option(uint8_t *data, uint16_t offset, uint16_t *pos, uint16_t max_len,
+			uint16_t *opt_delta, uint16_t *opt_len, struct coap_option *option);
```
Is this forward declaration needed? I don't see any new use of this function before the actual definition.

LukasWoodtli replied:
True. this is not needed anymore.

rlubos commented on subsys/net/lib/coap/coap.c:
```diff
@@ -49,53 +49,132 @@ LOG_MODULE_REGISTER(net_coap, CONFIG_COAP_LOG_LEVEL);
+		memmove(&cpkt->data[offset + 2], &cpkt->data[offset], cpkt->max_len - offset - 2);
+	}
+
+	cpkt->data[offset] = data >> 8;
+	cpkt->data[offset + 1] = (uint8_t)data;
+	cpkt->offset += 2;
```
This could also use `INSERT_BE16`?

LukasWoodtli replied:
True. I missed that one somehow.

rlubos commented on subsys/net/lib/coap/coap.c:
```diff
@@ -49,53 +49,132 @@ LOG_MODULE_REGISTER(net_coap, CONFIG_COAP_LOG_LEVEL);
+static int insert_option(struct coap_packet *cpkt, uint16_t code, const uint8_t *value,
+			 uint16_t len);
+
+#define INSERT_U8(cpkt, offst, data)                                                               \
```
I'm not really a fan of the name collision here, there already is a function called `insert_u8()`. Perhaps we could call it `encode_u8()`? And I'd prefer to make it a regular static inline function instead of a macro. Same for the macros below.

rlubos commented on subsys/net/lib/coap/coap.c:
```diff
@@ -266,16 +349,17 @@ int coap_packet_append_option(struct coap_packet *cpkt, uint16_t code,
-		NET_ERR("Options should be in ascending order");
-		return -EINVAL;
+		// This should be avoided. But it can be useful in some cases.
+		NET_INFO("Option is not added in ascending order");
```
Since out-of-order option processing is handled just fine, I'm not even sure if we need a log like this? Or at least make it a `DBG` level, IMO it doesn't need to be visiable at the default level (`INFO`).

rlubos commented on subsys/net/lib/coap/coap.c:
```diff
@@ -441,9 +525,8 @@ static int decode_delta(uint8_t *data, uint16_t offset, uint16_t *pos, uint16_t
-static int parse_option(uint8_t *data, uint16_t offset, uint16_t *pos,
-			uint16_t max_len, uint16_t *opt_delta, uint16_t *opt_len,
-			struct coap_option *option)
+int parse_option(uint8_t *data, uint16_t offset, uint16_t *pos, uint16_t max_len,
```
Why was the static removed?

LukasWoodtli replied:
The `static` was removed from the implementation as the declaration was added at the top of the file (with `static`) but since the declaration was not needed this will be reverted.

rlubos commented on subsys/net/lib/coap/coap.c:
```diff
@@ -981,6 +1064,55 @@ int coap_get_option_int(const struct coap_packet *cpkt, uint16_t code)
 	return val;
 }
 
+int insert_option(struct coap_packet *cpkt, uint16_t code, const uint8_t *value, uint16_t len)
```
This should be static as well?

LukasWoodtli replied:
The function is declared at the top of the file as `static`.

rlubos reviewed (commented):
Please take a look at the compliance results, there are quite a few failures like wrong comment type used.

rlubos commented:
There's also some mixup in the third commit: https://github.com/zephyrproject-rtos/zephyr/pull/54227/commits/f4a25044d10a27588bc88a17c4cf95cca20a72b0
As it touches both the library and tests. I think it'd make sense to stick to 2 commits only, one adding new feature, and second adding new tests. Please use `net: coap:` prefix title for implementation, and `tests: net: coap` for tests.

rlubos commented:
The commit message says:
> The insertion moves the data in the message buffer to make space for
the new option.
But it does not adjust the following options offset!

I don't follow `But it does not adjust the following options offset!` part, the option delta is updated, is there anything else to update I miss, or is it just some stale info?

LukasWoodtli commented:
> I don't follow `But it does not adjust the following options offset!` part, the option delta is updated, is there anything else to update I miss, or is it just some stale info?

Sorry, this is stale info.

I'll also clean up the commits.

LukasWoodtli commented:
> Please take a look at the compliance results, there are quite a few failures like wrong comment type used.

Unfortunately, I'm not able to run the checks on my machine. I get an exception in the Python script. But of course I'll check the results here on GitHub...

LukasWoodtli committed:
net: coap: Allow insertion of an option at arbitrary position

The insertion moves the data in the message buffer to make space for
the new option.

Signed-off-by: Lukas Woodtli <lukas.woodtli@husqvarnagroup.com>

LukasWoodtli committed:
tests: net: coap: Add test for options

The test checks if options are added to a CoAP buffer correctly.
It should eventually check if opstions can be added out of order
e.g. not in increasing order of option codes.

Signed-off-by: Lukas Woodtli <lukas.woodtli@husqvarnagroup.com>

rlubos commented on subsys/net/lib/coap/coap.c:
```diff
@@ -49,53 +49,122 @@ LOG_MODULE_REGISTER(net_coap, CONFIG_COAP_LOG_LEVEL);
 	return true;
 }
 
+
```
nit: Undeeded change.

rlubos commented on subsys/net/lib/coap/coap.c:
```diff
@@ -192,8 +261,11 @@ static uint8_t encode_extended_option(uint16_t num, uint8_t *opt, uint16_t *ext)
 
-static int encode_option(struct coap_packet *cpkt, uint16_t code,
-			 const uint8_t *value, uint16_t len)
+/* Insert an option at position `offset`. This is not adjusting the code delta of the
+ *  option that follows the inserted one!
```
```suggestion
/* Insert an option at position `offset`. This is not adjusting the code delta of the
 * option that follows the inserted one!
```

rlubos reviewed (commented):
Just some final nits, otherwise looks good for me.

pdgendt commented on subsys/net/lib/coap/coap.c:
```diff
@@ -49,49 +49,117 @@ LOG_MODULE_REGISTER(net_coap, CONFIG_COAP_LOG_LEVEL);
+}
+
+static bool enough_space(struct coap_packet *cpkt, const uint16_t bytes_to_add)
 {
 	if (!cpkt) {
 		return false;
 	}
 
-	if (cpkt->max_len - cpkt->offset < 1) {
+	if (cpkt->max_len - cpkt->offset < bytes_to_add) {
+		return false;
+	}
+
+	return true;
+}
```
This can be an inline oneliner.

```c
static inline bool enough_space(struct coap_packet *cpkt, const uint16_t bytes_to_add)
{
	return (cpkt != NULL) && (cpkt->max_len - cpkt->offset >= bytes_to_add);
}
```

LukasWoodtli replied:
imho it's less readable. But it might be more performant

pdgendt commented on subsys/net/lib/coap/coap.c:
```diff
@@ -49,49 +49,117 @@ LOG_MODULE_REGISTER(net_coap, CONFIG_COAP_LOG_LEVEL);
 static inline bool append(struct coap_packet *cpkt, const uint8_t *data, uint16_t len)
 {
-	if (!cpkt || !data) {
+	if (!enough_space(cpkt, len)) {
+		return false;
+	}
+	if (!data) {
 		return false;
 	}
```
Group some checks:

```c
if (data == NULL || !enough_space(cpkt, len)) {
	return false;
}
```
